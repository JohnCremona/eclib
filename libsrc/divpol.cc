// divpol.cc: implementations of functions for division polynomials
//////////////////////////////////////////////////////////////////////////
//
// Copyright 1990-2023 John Cremona
// 
// This file is part of the eclib package.
// 
// eclib is free software; you can redistribute it and/or modify it
// under the terms of the GNU General Public License as published by the
// Free Software Foundation; either version 2 of the License, or (at your
// option) any later version.
// 
// eclib is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
// FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
// for more details.
// 
// You should have received a copy of the GNU General Public License
// along with eclib; if not, write to the Free Software Foundation,
// Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA
// 
//////////////////////////////////////////////////////////////////////////
 
// NB the external interface uses a simple vector<bigint> of
// coefficients rather than a polynomial type to simplify the
// NTL interface

#include <eclib/polys.h>
#include <eclib/divpol.h>

// The 2-divison polynomial (cubic in x)

ZPoly div_pol_2(const bigint& a1,const bigint& a2,const bigint& a3,const bigint& a4,
                const bigint& a6)
{
  ZPoly ans;
  SetDegree(ans,3);
  SetCoeff(ans,3,4);
  SetCoeff(ans,2,a1*a1+4*a2);
  SetCoeff(ans,1,2*a1*a3+4*a4);
  SetCoeff(ans,0,a3*a3+4*a6);
  return ans;
}

// div_pol_odd(a1,a2,a3,a4,a6,n) returns the coefficients of the
// polynomial in x whose zeros are the (x-coordinates of the) non-zero
// points P on E=[a1,a2,a3,a4,a6] satisfying nP=0 (odd n)

// The poly itself is found recursively

ZPoly div_pol_odd(const bigint& a1,const bigint& a2,const bigint& a3,const bigint& a4,
                  const bigint& a6, int n)
{
  static const bigint four(4);
  ZPoly X; ZPolySetX(X);
  ZPoly f1 = X*(X*(X+a2)+a4)+a6;
  ZPoly f2 = a1*X+a3;
  ZPoly psi24=(four*f1+f2*f2); psi24*=psi24;
  ZPoly ans;
  switch(n) {
  case 0: 
    SetDegree(ans,0); SetCoeff(ans,0,0); return ans;
  case 1: case 2: 
    SetDegree(ans,0); SetCoeff(ans,0,1); return ans;
  case 3:
    SetDegree(ans,4);
    SetCoeff(ans,4,3);
    SetCoeff(ans,3,a1*a1+4*a2);
    SetCoeff(ans,2,3*a1*a3+6*a4);
    SetCoeff(ans,1,3*a3*a3+12*a6);
    SetCoeff(ans,0,a1*a1*a6-a1*a3*a4+a2*a3*a3+4*a2*a6-a4*a4);
    return ans;
  case 4:
    SetDegree(ans,6);
    SetCoeff(ans,6,2);
    SetCoeff(ans,5,a1*a1+4*a2);
    SetCoeff(ans,4,5*a1*a3+10*a4);
    SetCoeff(ans,3,10*a3*a3+40*a6);
    SetCoeff(ans,2,10*a1*a1*a6-10*a1*a3*a4+10*a2*a3*a3+40*a2*a6-10*a4*a4);
    SetCoeff(ans,1,a1*a1*a1*a1*a6-a1*a1*a1*a3*a4+a1*a1*a2*a3*a3+8*a1*a1*a2*a6-a1*a1*a4*a4-4*a1*a2*a3*a4-a1*a3*a3*a3-4*a1*a3*a6+4*a2*a2*a3*a3+16*a2*a2*a6-4*a2*a4*a4-2*a3*a3*a4-8*a4*a6);
    SetCoeff(ans,0,a1*a1*a1*a3*a6-a1*a1*a3*a3*a4+2*a1*a1*a4*a6+a1*a2*a3*a3*a3+4*a1*a2*a3*a6-3*a1*a3*a4*a4+2*a2*a3*a3*a4+8*a2*a4*a6-a3*a3*a3*a3-8*a3*a3*a6-2*a4*a4*a4-16*a6*a6);
    return ans;
  default: // general case, use recursion
    // If n is odd, n=2m+1:
    if(n%2==1)
      {
	int m=(n-1)/2;
	ZPoly t1=div_pol_odd(a1,a2,a3,a4,a6,m);
	t1=div_pol_odd(a1,a2,a3,a4,a6,m+2)*t1*t1*t1;
	ZPoly t2=div_pol_odd(a1,a2,a3,a4,a6,m+1);
	t2=div_pol_odd(a1,a2,a3,a4,a6,m-1)*t2*t2*t2;
	if(m%2==1) return t1-psi24*t2;
	return psi24*t1-t2;
      }
    else // n is even, n=2m:
      {
	int m=n/2;
	ZPoly t1=div_pol_odd(a1,a2,a3,a4,a6,m-1);
	t1=div_pol_odd(a1,a2,a3,a4,a6,m+2)*t1*t1;
	ZPoly t2=div_pol_odd(a1,a2,a3,a4,a6,m+1);
	t2=div_pol_odd(a1,a2,a3,a4,a6,m-2)*t2*t2;
	return div_pol_odd(a1,a2,a3,a4,a6,m)*(t1-t2);
      }
  }
}

ZPoly div_pol(const bigint& a1,const bigint& a2,const bigint& a3,const bigint& a4,
              const bigint& a6,int n)
{
  return (n==2? div_pol_2(a1,a2,a3,a4,a6) :  div_pol_odd(a1,a2,a3,a4,a6,n));
}

ZPoly division_polynomial(Curvedata* E, int p)
{
  bigint a1,a2,a3,a4,a6;
  E->getai(a1,a2,a3,a4,a6);
  return (p==2? div_pol_2(a1,a2,a3,a4,a6) :  div_pol_odd(a1,a2,a3,a4,a6,p));
}

// Numerator of the multiplication-by-n map on the x-coordinate

ZPoly mul_by_n_num(const bigint& a1,const bigint& a2,const bigint& a3,const bigint& a4,
                   const bigint& a6, int n)
{
  ZPoly X; ZPolySetX(X);
  ZPoly P_2 = div_pol_2(a1,a2,a3,a4,a6);
  ZPoly P_n = div_pol_odd(a1,a2,a3,a4,a6,n);
  ZPoly P_nplus1 = div_pol_odd(a1,a2,a3,a4,a6,n+1);
  ZPoly P_nminus1 = div_pol_odd(a1,a2,a3,a4,a6,n-1);
  ZPoly A = X * P_n * P_n;
  ZPoly B = P_nminus1 * P_nplus1;
  return (n%2==0? A * P_2 - B : A - P_2 * B);
}

// Denominator of the multiplication-by-n map on the x-coordinate

ZPoly mul_by_n_den(const bigint& a1,const bigint& a2,const bigint& a3,const bigint& a4,
                   const bigint& a6, int n)
{
  ZPoly P_n = div_pol_odd(a1,a2,a3,a4,a6,n);
  return (n%2==1? P_n * P_n : P_n * P_n * div_pol_2(a1,a2,a3,a4,a6));
}

// Polynomial whose roots are x(Q) for Q satisfying n*Q=P, where x(P)=xP/zP

ZPoly division_points_X_pol(const bigint& a1,const bigint& a2,const bigint& a3,const bigint& a4,
                            const bigint& a6,
                            int n,
                            const bigint& xP, const bigint& zP)
{
  ZPoly numpoly = mul_by_n_num(a1, a2, a3, a4, a6, n);
  ZPoly denpoly = mul_by_n_den(a1, a2, a3, a4, a6, n);
  //cout << "numpoly = " << numpoly << endl;
  //cout << "denpoly = " << numpoly << endl;
  return zP * numpoly - xP * denpoly;
}
