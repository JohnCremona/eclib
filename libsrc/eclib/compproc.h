// compproc.h: declarations of functions using complex numbers
//////////////////////////////////////////////////////////////////////////
//
// Copyright 1990-2023 John Cremona
// 
// This file is part of the eclib package.
// 
// eclib is free software; you can redistribute it and/or modify it
// under the terms of the GNU General Public License as published by the
// Free Software Foundation; either version 2 of the License, or (at your
// option) any later version.
// 
// eclib is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
// FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
// for more details.
// 
// You should have received a copy of the GNU General Public License
// along with eclib; if not, write to the Free Software Foundation,
// Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA
// 
//////////////////////////////////////////////////////////////////////////
 
#ifndef _ECLIB_COMPPROC_H_
#define _ECLIB_COMPPROC_H_

#include <eclib/interface.h>

int is_small(bigfloat x);
int is_small(const bigcomplex& z);
int is_real(const bigcomplex& z);

void orderreal(bigfloat& e1, bigfloat& e2, bigfloat& e3);  // puts in decreasing order
bigcomplex root(const bigcomplex& z, int n);
bigcomplex cagm(const bigcomplex& a, const bigcomplex& b);
bigcomplex normalize(bigcomplex& w1, bigcomplex& w2);
void getc4c6(const bigcomplex& w1, const bigcomplex& w2, bigcomplex& c4, bigcomplex &c6);
bigcomplex discriminant(const bigcomplex& b, const bigcomplex& c, const bigcomplex& d);

vector<bigcomplex> solvecubic(const bigcomplex& c1, const bigcomplex& c2, const bigcomplex& c3);
inline vector<bigcomplex> solvecubic(const bigfloat& c1, const bigfloat& c2, const bigfloat& c3)
{return solvecubic(bigcomplex(c1), bigcomplex(c2), bigcomplex(c3));}

vector<bigcomplex> solvequartic(const bigcomplex& a, const bigcomplex& b, const bigcomplex& c, const bigcomplex& d);
inline vector<bigcomplex> solvequartic(const bigfloat& a, const bigfloat& b, const bigfloat& c, const bigfloat& d)
{return solvequartic(bigcomplex(a), bigcomplex(b), bigcomplex(c), bigcomplex(d));}

vector<bigcomplex> solverealquartic(const bigfloat& a, const bigfloat& b, const bigfloat& c, const bigfloat& d, const bigfloat& e);

void quadsolve(const bigfloat& p, const bigfloat& q, bigcomplex& root1,bigcomplex& root2);

#endif
