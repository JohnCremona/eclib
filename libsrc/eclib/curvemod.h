// curvemod.h: declaration of class curvemodq for elliptic curve mod (prime) q 
//////////////////////////////////////////////////////////////////////////
//
// Copyright 1990-2012 John Cremona
// 
// This file is part of the eclib package.
// 
// eclib is free software; you can redistribute it and/or modify it
// under the terms of the GNU General Public License as published by the
// Free Software Foundation; either version 2 of the License, or (at your
// option) any later version.
// 
// eclib is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
// FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
// for more details.
// 
// You should have received a copy of the GNU General Public License
// along with eclib; if not, write to the Free Software Foundation,
// Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA
// 
//////////////////////////////////////////////////////////////////////////
 

// allow for multiple includes
#ifndef _ECLIB_CURVEMOD_
#define _ECLIB_CURVEMOD_

#include <eclib/polys.h>
#include <eclib/curve.h>

// Class for an elliptic curve mod q
// 
// The q is fixed -- must be set before use!
//


class pointmodq;

class curvemodq  { 

protected:
  galois_field* Fq;           // pointer to ground field
  bigint q;                   //  the modulus
  gf_element a1,a2,a3,a4,a6;  //  the coefficients mod q
  bigint order;               // number of points (0 if not set)

public:
  // constructors 
  curvemodq(void);
  curvemodq(const Curve& E, const bigint& qq);
  ~curvemodq();
  curvemodq(const curvemodq& C); // copy constructor
  void operator=(const curvemodq& C); // assignment
  
  // access
  void get_ai(gf_element& aa1, gf_element& aa2, gf_element& aa3, gf_element& aa4, gf_element& aa6) const
    {
      aa1=a1; aa2=a2; aa3=a3; aa4=a4; aa6=a6;
    }

  // output
  void output(ostream& os) const {os<<"["<<a1<<","<<a2<<","<<a3<<","<<a4<<","<<a6<<"] mod "<<q;}
  
  // equality test
  int operator==(const curvemodq& C) const
    { return (q==C.q)&&(a1==C.a1)&&(a2==C.a2)&&(a3==C.a3)&&(a4==C.a4)&&(a6==C.a6);}
  int operator!=(const curvemodq& C) const
    { return (q!=C.q)||(a1!=C.a1)||(a2!=C.a2)||(a3!=C.a3)||(a4!=C.a4)||(a6!=C.a6);}

  // random point
  pointmodq random_point(); // defined in pointsmod.cc

  bigint get_modulus() const { return q; }

  void set_group_order_via_legendre(); 
  void set_group_order();
  bigint group_order() {if(is_zero(order)) set_group_order(); return order;}
  
  friend class galois_field;
  friend class ffmodq;
  friend class pointmodq;
  friend class TLSS;
  friend galois_field get_field(const curvemodq& C);
  friend bigint get_modulus(const curvemodq& E);
  };

inline ostream& operator<<(ostream& os, const curvemodq& C)
{
  C.output(os);
  return os;
}

  // just a wrapper round the constructor, (originally for for LiDIA-compatibility):

inline curvemodq reduce_curve(const Curve& E, const bigint& q)
{
  return curvemodq(E,q);
}

inline galois_field get_field(const curvemodq& C) { return galois_field(C.q); }
inline bigint get_modulus(const curvemodq& C) { return C.q; }

FqPoly div_pol_odd(const curvemodq& C, int n); 
FqPoly makepdivpol(const curvemodq& C, int p);


#endif // #define _CURVEMOD_
