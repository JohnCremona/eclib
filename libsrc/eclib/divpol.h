// divpol.h: declaration of functions for division polynomials
//////////////////////////////////////////////////////////////////////////
//
// Copyright 1990-2023 John Cremona
// 
// This file is part of the eclib package.
// 
// eclib is free software; you can redistribute it and/or modify it
// under the terms of the GNU General Public License as published by the
// Free Software Foundation; either version 2 of the License, or (at your
// option) any later version.
// 
// eclib is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
// FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
// for more details.
// 
// You should have received a copy of the GNU General Public License
// along with eclib; if not, write to the Free Software Foundation,
// Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA
// 
//////////////////////////////////////////////////////////////////////////
 
// allow for multiple includes
#ifndef _ECLIB_DIVPOL_
#define _ECLIB_DIVPOL_

#include "curve.h"

// These functions return the vector of coefficients (starting with
// the constant term) rather than polynomial types

// div_pol_odd(a1,a2,a3,a4,a6,n) returns the coefficients of the
// polynomial in x whose zeros are the (x-coordinates of the) non-zero
// points P on E=[a1,a2,a3,a4,a6] satisfying nP=0 (odd n)

// The poly itself is found recursively

// Despite the name, for even n this returns a correct n-division
// polynomial without the 2-torsion factor, i.e. the polynomial whoe
// roots are the x-coordinates of the points P satisfying nP=0, 2P!=0.

ZPoly div_pol_odd(const ZZ& a1,const ZZ& a2,const ZZ& a3,const ZZ& a4,
                  const ZZ& a6,int n);

ZPoly div_pol_2(const ZZ& a1,const ZZ& a2,const ZZ& a3,const ZZ& a4,
                const ZZ& a6);

ZPoly div_pol(const ZZ& a1,const ZZ& a2,const ZZ& a3,const ZZ& a4,
              const ZZ& a6,int n);

ZPoly division_polynomial(Curvedata* EE, int p);

// Numerator and denominator of the multiplication-by-n map on the x-coordinate

ZPoly mul_by_n_num(const ZZ& a1,const ZZ& a2,const ZZ& a3,const ZZ& a4,
                   const ZZ& a6, int n);

ZPoly mul_by_n_den(const ZZ& a1,const ZZ& a2,const ZZ& a3,const ZZ& a4,
                   const ZZ& a6, int n);

// Polynomial whose roots are x(Q) for Q satisfying n*Q=P, where x(P)=xP/zP

ZPoly division_points_X_pol(const ZZ& a1,const ZZ& a2,const ZZ& a3,const ZZ& a4,
                            const ZZ& a6,
                            int n,
                            const ZZ& xP, const ZZ& zP);

#endif // #define _DIVPOL_
