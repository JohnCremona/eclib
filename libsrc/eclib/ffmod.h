// ffmod.h: declaration of class ffmodq and Weil pairing functions
//////////////////////////////////////////////////////////////////////////
//
// Copyright 1990-2023 John Cremona
// 
// This file is part of the eclib package.
// 
// eclib is free software; you can redistribute it and/or modify it
// under the terms of the GNU General Public License as published by the
// Free Software Foundation; either version 2 of the License, or (at your
// option) any later version.
// 
// eclib is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
// FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
// for more details.
// 
// You should have received a copy of the GNU General Public License
// along with eclib; if not, write to the Free Software Foundation,
// Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA
// 
//////////////////////////////////////////////////////////////////////////
 

// ffmodq is the function field of an elliptic curve mod a prime q
// (or more precisely the affine coordinate ring Fq[x,y])

// allow for multiple includes
#ifndef _ECLIB_FFMOD_
#define _ECLIB_FFMOD_

#include "pointsmod.h"

class ffmodq{

  public:

  // static members defining the function field:
  static galois_field Fq;        //  the constant field
  static curvemodq E;            //  the curve mod q
  static ZZ_pX f1, f2;          //  f2=a1*x+a3, f2=x^3+a2*x^2+a4*x+a6
  static void init(const curvemodq& EE); // set Fq, E, f1, f2

  // data defining one element of th function field:
  ZZ_pX h1, h2;                 //  for h1+y*h2


public:

  //  constructors

  //  normal ones:
  ffmodq(void)
    {
      init_h1h2();
      h1 = to_ZZ_p(0);
      h2 = to_ZZ_p(0);
    }

  explicit ffmodq(const ZZ_p& c)
    {
      init_h1h2();
      h1 = c;
      h2 = to_ZZ_p(0);
    }

  explicit ffmodq(const ZZ& c)
    {
      init_h1h2();
      h1 = to_ZZ_p(c);
      h2 = to_ZZ_p(0);
    }

  explicit ffmodq(const ZZ_pX& hh1)
    {
      init_h1h2();
      h1 = hh1;
      h2 = to_ZZ_p(0);
    }

  ffmodq(const ZZ_pX& hh1, const ZZ_pX& hh2) :h1(hh1), h2(hh2) {}

  //  initialization
  void init_h1h2(void)
    {
      // ZZ_pXSetField(h1,Fq);
      // ZZ_pXSetField(h2,Fq);
    }

  // equality test
  int operator==(const ffmodq& b) const;
  int operator!=(const ffmodq& b) const {return !((*this)==b);}

  // output
  void output(ostream& os) const;

  // addition, subtraction, multiplication
  ffmodq operator+(const ffmodq& b) const;
  ffmodq operator-(const ffmodq& b) const;
  ffmodq operator*(const ffmodq& b) const;
  ffmodq operator*(const ZZ_pX& h) const;

  //  division
  ffmodq operator/(const ZZ_pX& h) const;
  ffmodq operator/(const ffmodq& b) const;

  //  evaluation at a point:
  ZZ_p evaluate(const pointmodq& P) const;
  ZZ_p operator()(const pointmodq& P) const {return this->evaluate(P);}

  //  vertical line through a point:
  friend ffmodq vertical(const pointmodq& P);

  //  tangent at a point:
  friend ffmodq tangent(const pointmodq& P);

  //  chord between points:
  friend ffmodq chord(const pointmodq& P, const pointmodq& Q);


};

// weil_pol(T,m): T is a point of finite order m; returns a function
// f_T whose divisor is m(T)-m(0).

// The second version evaluates that at another point S without
// actually computing the polynomial

ffmodq weil_pol(const pointmodq& T, int m);

ZZ_p evaluate_weil_pol(const pointmodq& T, int m, const pointmodq& S);

ZZ_p weil_pairing(const pointmodq& S, const pointmodq& T, int m);

inline ostream& operator<<(ostream& os, const ffmodq& f)
{
  f.output(os);
  return os;
}


#endif
