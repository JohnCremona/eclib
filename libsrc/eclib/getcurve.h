// getcurve.h: declaration of function getcurve() for curve input
//////////////////////////////////////////////////////////////////////////
//
// Copyright 1990-2012 John Cremona
// 
// This file is part of the eclib package.
// 
// eclib is free software; you can redistribute it and/or modify it
// under the terms of the GNU General Public License as published by the
// Free Software Foundation; either version 2 of the License, or (at your
// option) any later version.
// 
// eclib is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
// FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
// for more details.
// 
// You should have received a copy of the GNU General Public License
// along with eclib; if not, write to the Free Software Foundation,
// Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA
// 
//////////////////////////////////////////////////////////////////////////

#if     !defined(_ECLIB_GETCURVE_H)
#define _ECLIB_GETCURVE_H      1       //flags that this file has been included

#include <eclib/curve.h>

// Read in a curve as [a1,a2,a3,a4,a6] with ai integers:
int getcurve(Curvedata& CD, int verb);

// Read in a curve as [a1,a2,a3,a4,a6] with ai rational:
int getcurve(vector<bigrational>& ai, int verb);

#endif
