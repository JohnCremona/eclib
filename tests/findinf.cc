// findinf.cc:  program to find points up to given naive height
//////////////////////////////////////////////////////////////////////////
//
// Copyright 1990-2005 John Cremona
// 
// This file is part of the mwrank package.
// 
// mwrank is free software; you can redistribute it and/or modify it
// under the terms of the GNU General Public License as published by the
// Free Software Foundation; either version 2 of the License, or (at your
// option) any later version.
// 
// mwrank is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
// FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
// for more details.
// 
// You should have received a copy of the GNU General Public License
// along with mwrank; if not, write to the Free Software Foundation,
// Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
// 
//////////////////////////////////////////////////////////////////////////
 
#include "interface.h"
#include "compproc.h"

#include "matrix.h"
#include "subspace.h"

#include "points.h"
#include "polys.h"
#include "curvemod.h"
#include "pointsmod.h"
#include "ffmod.h"
#include "divpol.h"
#include "tlss.h"
#include "elog.h"
#include "saturate.h"
#include "getcurve.h"
#include "sieve_search.h"
#include "mwprocs.h"

int main()
{
  set_precision(string("Enter number of decimal places").c_str());
  initprimes(string("PRIMES").c_str(),0);

  bigfloat ht_limit;
  bigint u,r,s,t;
  int verbose = 1, modopt=0, pp=1, change_flag; 
  long blength, rank, maxrank;
  cerr<<"\nenter search limit: ";      cin>>ht_limit;
  cerr<<"verbose (0/1)? ";             cin >>verbose;
  cerr<<"process points found (1) or just list them (0): "; cin>>pp;
  //  cout<<"moduli option (0 (Stoll)/ 1/2/3)?";      cin >> modopt;
  int verb=1;
  bigint v;
  vector<bigrational> ai(5);

  while (getcurve(ai,verb))
    {
      Curvedata C(ai,v);
      cout << "Input curve ";
      cout <<"["<<ai[0]<<","<<ai[1]<<","<<ai[2]<<","<<ai[3]<<","<<ai[4]<<"]" << endl;
      Curvedata C_min = C.minimalize(u,r,s,t);
      change_flag = (v!=1) || (((Curve)C_min) != (Curve)C);
      if(change_flag)
	{
	  cout<<"Searching on standard minimal model "<<(Curve)C_min<<endl;
	  cout<<"(points found will be transferred back at end)"<<endl;
	  cout<<"Transformation: \t[u,r,s,t] = ["<<u<<","<<r<<","<<s<<","<<t<<"]";
	  if(v!=1) cout<<" with scale factor "<<v;
	  cout<<endl;
	}
      
      P2Point P0;
      cerr<<"enter number of known points: ";      cin >>blength;
      cerr<<"enter max rank to stop when this is reached (-ve for none): ";
      cin >>maxrank; if(maxrank<0) maxrank=999;
      vector<Point> known_points(blength);
      if (blength)
	{ 
	  for (long j=0; j<blength; j++)
	    { cerr<<"\n  enter point "<<(j+1); 
	      if(change_flag) cerr<<" on original curve";
	      cerr<<": ";
	      cin >> P0;
	      Point P(C,scale(P0,v,1));
	      if ( P.isvalid() ) 
		{
		  Point Q = transform(P,&C_min,u,r,s,t,0);
		  known_points[j] = Q;
		  if(change_flag) 
		    {
		      cout<<P0<<" maps to "<<Q<<" on "<<(Curve)C_min<<", with height "<<height(Q)<<endl;
		    }
		}
	      else 
		{
		  cout<<"Bad point -- not on curve. Start all over again.\n";
		  break; 
		}
	    }
	}

      mw mwbasis(&C_min, verbose, pp, maxrank);
      if(blength)
	{
	  mwbasis.process(known_points);
	  rank = mwbasis.getrank();
	  cout<<"Rank of known points is "<<rank
	      <<" with regulator "<<mwbasis.regulator()<<endl<<endl;
	}

      mwbasis.search(ht_limit, modopt, verbose);
      
      if(pp)
	{
	  bigint index;
	  vector<long> unsat;
	  if(verbose) 
	    {
	      cout <<"Regulator (before saturation) = "<<mwbasis.regulator()<<"\n";
	      cout <<"Saturating..." << flush;
	    }
	  int sat_ok = mwbasis.saturate(index,unsat);
	  if(verbose)
	    {
	      cout <<"finished saturation (index was " << index << ")\n";
	      cout <<"Regulator (after saturation) = "<<mwbasis.regulator()<<"\n";
	    }
	  if(!sat_ok) 
	    {
	      cout << "saturation possibly incomplete at primes " << unsat << "\n";
	    }
	  rank = mwbasis.getrank();
	  cout<<"Rank of points found is "<<rank<<endl;
	  vector<Point> b = mwbasis.getbasis();
	  vector<P2Point> bb(rank);
	  for (long i=0; i<rank; i++)
	    { 
	      Point Q = b[i];
	      bigfloat hQ = height(Q);
	      cout << "Generator "<<(i+1)<<" is "<<Q<<"; ";
	      cout << "height "<<hQ<<endl;
	      P0 = scale(transform(Q,&C,u,r,s,t,1),v,0);
	      bb[i] = P0;
	      if(change_flag)
		cout<<"\t--maps back to "<<P0<<" on input curve"<<endl;
	    }
	  if(rank>1) cout<<"Regulator =  "<<mwbasis.regulator()<<endl;
	  if(sat_ok)
	    {
	      cout<<"Points have been successfully saturated"<<endl;
	    }
	  else
	    {
	      cout<<"Saturation incomplete "<<endl;
	    }
	  cout<<endl;
// Pari output:
	  cout<<"[";
	  for(long i=0; i<rank; i++)
	    {
	      if(i) cout<<",";
	      output_pari(cout,bb[i]);
	    }
	  cout<<"]\n"<<endl;
	}
    }
  cout<<endl;
}

//end of file findinf.cc





