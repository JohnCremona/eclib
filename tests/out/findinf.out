Input curve [0,0,1,-7,6]
Rank of points found is 3
Generator 1 is [1:-1:1]; height 0.66820516565192793502
Generator 2 is [-2:3:1]; height 1.368572505353930112
Generator 3 is [-14:25:8]; height 2.7173593928122930896
Regulator =  0.41714355875838397005
Points have been successfully saturated

[[1,-1],[-2,3],[-7/4,25/8]]

Input curve [0,1,1,-2,0]
Rank of points found is 2
Generator 1 is [0:-1:1]; height 0.32700077365160495184
Generator 2 is [-1:1:1]; height 0.68666708330558658573
Regulator =  0.15246017794314375163
Points have been successfully saturated

[[0,-1],[-1,1]]

Input curve [2/5,-1/45,1/1800,-7/810000,7/60750000]
Searching on standard minimal model [0,1,1,-2,0]
(points found will be transferred back at end)
Transformation: 	[u,r,s,t] = [1,-5,-6,23] with scale factor 30
[-6:1:900] maps to [-1:1:1] on [0,1,1,-2,0], with height 0.68666708330558658573
Rank of known points is 1 with regulator 0.68666708330558658573

Rank of points found is 2
Generator 1 is [0:-1:1]; height 0.32700077365160495184
	--maps back to [-75:11:13500] on input curve
Generator 2 is [-2:0:1]; height 0.92075778268510239274
	--maps back to [-42:7:5400] on input curve
Regulator =  0.15246017794314375163
Points have been successfully saturated

[[-1/180,11/13500],[-7/900,7/5400]]


